### PR TITLE
[CUDA] Add GatedLatentPool and SparseRowSelect contrib operators

### PR DESCRIPTION
**Pull request by tianleiwu**
## Description

Adds two CUDA contrib operators that together form the compression and selection front end of a sparse-attention stack: `GatedLatentPool` pools `ratio` consecutive tokens into one latent cache row, and `SparseRowSelect` scores those rows against the current query and emits the `topk` row ids attention should gather. DeepSeek-V4 uses this pair (its KV compressor and its Lightning Indexer), but the schemas are written against the general pattern — gated pooling into a strided latent cache, then a lightweight multi-head scorer over that cache.

As ONNX primitives each of these is a long chain of gather/softmax/reduce/rotate nodes whose intermediates are far larger than their inputs; fusing them removes both the launches and the materialization.

### `GatedLatentPool`

| File | Change |
|------|--------|
| `onnxruntime/contrib_ops/cuda/math/gated_latent_pool.{h,cc}` | `OpKernel` — attributes, validation, output shapes |
| `onnxruntime/contrib_ops/cuda/math/gated_latent_pool_impl.{h,cu}` | Fused pooling, RMS norm, rotary and optional quantization-simulation kernels |

Gated pooling of `ratio` consecutive tokens into one latent row:

```
w      = Softmax(Where(valid, full_score[window] + ape, -1e30), over the window)
pooled = ReduceSum(full_kv[window] * w, over the window)
normed = norm_weight * pooled / Sqrt(ReduceMean(pooled * pooled, -1) + epsilon)
```

Design points worth a reviewer's attention:

- The operator carries a **rolling state** of the `window_multiplier * ratio` projections that preceded this step, so a row whose window straddles a step boundary can still be pooled without the caller re-feeding history.
- `score` is optional: when omitted, `kv` is twice as wide and carries the value in the low half of each row and the gate in the high half, which is what a single fused projection GEMM produces.
- `rows` is sized `(seq_len - 1) / ratio + 2` — one row per slot the step can touch plus a spare — so **the output shape never depends on `past_lens`**, which keeps the op CUDA-graph-capturable. `first_slot` / `last_slot` / `row_count` say which of those rows are live.
- `simulate_fp8` and `simulate_rotated_fp4` apply a simulated FP8-E4M3 / Hadamard-rotated FP4-E2M1 round trip, matching what a quantized attention cache will store.

### `SparseRowSelect`

| File | Change |
|------|--------|
| `onnxruntime/contrib_ops/cuda/math/sparse_row_select.{h,cc}` | `OpKernel` — attributes, validation, output shapes |
| `onnxruntime/contrib_ops/cuda/math/sparse_row_select_impl.{h,cu}` | Rotary + optional FP4 simulation, cache update, scoring and top-k kernels |

```
score[b, s, c] = sum_h Relu(dot(query[b, s, h], present_cache[b, c])) * weights[b, s, h] * scale
```

- Rows produced by `GatedLatentPool` this step are written at `first_slot + j`; slots below keep `past_cache`. **Slots above `last_slot` are deliberately left unspecified** — no step has ever written them and no query can reach them. On a long-context export they are almost the whole cache, so producing them would dominate the operator.
- A query at absolute position `past_lens[b] + s` may only see the first `(past_lens[b] + s + 1) / ratio` rows.
- `selection` is emitted as `row_id_offset + c` so it can be concatenated straight into a paged cache's index list, padded with `-1` when fewer than `topk` rows are visible.
- Attention does not depend on gather order, so the selection is emitted in **ascending row order** rather than by descending score — this avoids a sort and makes the output deterministic.

### Shared

| File | Change |
|------|--------|
| `onnxruntime/contrib_ops/cuda/math/quant_sim_common.cuh` | Shared quantization-simulation helpers (`QuantSimConv`, block scale, E4M3 rounding, FP4 rotation) used by both operators |
| `onnxruntime/core/graph/contrib_ops/contrib_defs.cc` | Schemas, shape inference and the formulas above |
| `onnxruntime/core/graph/contrib_ops/ms_opset.h` | Schema registration |
| `onnxruntime/contrib_ops/cuda/cuda_contrib_kernels.cc` | CUDA kernel registration for float, float16 and bfloat16 |
| `docs/ContribOperators.md`, `docs/OperatorKernels.md` | Regenerated entries |

## Testing

- `onnxruntime/test/python/transformers/test_gated_latent_pool.py`
- `onnxruntime/test/python/transformers/test_sparse_row_select.py`

Both compare against PyTorch references built from the primitive formulas, across float32/float16/bfloat16, with and without the optional `score` input, with and without each quantization-simulation mode, and across window multipliers, ratios, and prefill/decode step shapes.

```
python -m pytest onnxruntime/test/python/transformers/test_gated_latent_pool.py \
                 onnxruntime/test/python/transformers/test_sparse_row_select.py
# 11 passed, 138 subtests passed
```

Additive only — no existing operator, schema or kernel is modified.

## Checklist

- [x] Tests added
- [x] No breaking changes
- [x] Documentation updated

### CODE REVIEW

File: onnxruntime/test/python/transformers/test_sparse_row_select.py [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+gather list can depend on.
+"""
+
+# Tensor dimensions keep the upper-case names the schema doc gives them.
+# ruff: noqa: N806
+
+import unittest
+
+import numpy as np
+import onnx
```
[the remark this answers is not in the document]
Comment (tianleiwu) [reply onnxruntime/test/python/transformers/test_sparse_row_select.py]: Fixed in 7257b0d210. Removed mixed `import onnx` and used `TensorProto.DataType.Name`. lintrunner passes.

File: onnxruntime/contrib_ops/cuda/math/gated_latent_pool.cc [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+  max_seq_len_ = info.GetAttrOrDefault<int64_t>("max_seq_len", static_cast<int64_t>(0));
+  epsilon_ = info.GetAttrOrDefault<float>("epsilon", 1e-6f);
+  simulate_fp8_ = info.GetAttrOrDefault<int64_t>("simulate_fp8", static_cast<int64_t>(0)) != 0;
+  simulate_rotated_fp4_ = info.GetAttrOrDefault<int64_t>("simulate_rotated_fp4", static_cast<int64_t>(0)) != 0;
+
+  ORT_ENFORCE(ratio_ >= 1, "ratio must be positive, got ", ratio_);
+  ORT_ENFORCE(window_multiplier_ == 1 || window_multiplier_ == 2, "window_multiplier must be 1 or 2, got ", window_multiplier_);
+  ORT_ENFORCE(head_dim_ > 0, "head_dim must be positive, got ", head_dim_);
+  ORT_ENFORCE(rope_head_dim_ >= 0 && rope_head_dim_ <= head_dim_,
+              "rope_head_dim must be in [0, head_dim], got ", rope_head_dim_);
```
[the remark this answers is not in the document]
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cuda/math/gated_latent_pool.cc]: Fixed in 7257b0d210. GatedLatentPool now rejects odd rope_head_dim; added constructor regression test; lintrunner/plugin build pass.

File: onnxruntime/contrib_ops/cuda/math/sparse_row_select.cc [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+  ratio_ = info.GetAttrOrDefault<int64_t>("ratio", static_cast<int64_t>(1));
+  topk_ = info.GetAttrOrDefault<int64_t>("topk", static_cast<int64_t>(0));
+  row_id_offset_ = info.GetAttrOrDefault<int64_t>("row_id_offset", static_cast<int64_t>(0));
+  scale_ = info.GetAttrOrDefault<float>("scale", 1.0f);
+  simulate_rotated_fp4_ = info.GetAttrOrDefault<int64_t>("simulate_rotated_fp4", static_cast<int64_t>(1)) != 0;
+
+  ORT_ENFORCE(num_heads_ > 0, "num_heads must be positive, got ", num_heads_);
+  ORT_ENFORCE(head_dim_ > 0, "head_dim must be positive, got ", head_dim_);
+  ORT_ENFORCE(rope_head_dim_ >= 0 && rope_head_dim_ <= head_dim_,
+              "rope_head_dim must be in [0, head_dim], got ", rope_head_dim_);
```
[the remark this answers is not in the document]
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cuda/math/sparse_row_select.cc]: Fixed in 7257b0d210. SparseRowSelect now rejects odd rope_head_dim; added constructor regression test; lintrunner/plugin build pass.

File: onnxruntime/contrib_ops/cuda/math/sparse_row_select_impl.cu [hunk 4]
```diff
@@ -0,0 +1,667 @@
...
+  // exact, 0 disables it, 2 forces it wherever it is legal -- which is how the parity test
+  // reaches it, since a test never captures a graph and so always has an exact bound.
+  static const int mma_mode =
+      ParseEnvironmentVariableWithDefault<int>("ORT_LIGHTNING_INDEXER_MMA_SCORE", 1);
+  const size_t mma_shared = static_cast<size_t>(SparseRowSelectScoreMmaSharedBytes(p));
+  const bool mma_score = mma_mode > 0 && (mma_mode > 1 || !p.score_capacity_exact) &&
+                         std::is_same<T, BFloat16>::value && p.simulate_rotated_fp4 &&
+                         p.seq_len <= kMaxFusedScoreSeq && p.num_heads % kMmaTile == 0 &&
+                         p.head_dim % kMmaTile == 0 &&
+                         mma_shared <= static_cast<size_t>(prop.sharedMemPerBlockOptin);
```
[the remark this answers is not in the document]
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cuda/math/sparse_row_select_impl.cu]: Fixed in 7257b0d210. WMMA dispatch requires prop.major >= 8, so older devices fall back; plugin build passes.

File: onnxruntime/contrib_ops/cuda/math/sparse_row_select.cc [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+  params.head_dim = static_cast<int>(head_dim_);
+  params.rope_head_dim = static_cast<int>(rope_head_dim_);
+  params.nope_dim = static_cast<int>(head_dim_ - rope_head_dim_);
+  params.num_rows = static_cast<int>(num_rows);
+  params.capacity = static_cast<int>(capacity);
+  params.score_capacity = static_cast<int>(capacity);
+  params.score_capacity_exact = false;
+  params.ratio = static_cast<int>(ratio_);
+  params.topk = static_cast<int>(topk_);
+  params.row_id_offset = static_cast<int>(row_id_offset_);
```
[the remark this answers is not in the document]
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cuda/math/sparse_row_select.cc:200]: Fixed in 7257b0d210. Validate row_id_offset <= numeric_limits<int>::max() before narrowing and added an oversized-value constructor regression test; the plugin build passes.

File: onnxruntime/test/python/transformers/test_sparse_row_select.py [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+        )
+
+    def test_small_no_rotation(self):
+        self._run_case(
+            {"nh": 4, "hd": 32, "rd": 16, "ratio": 4, "L": 64, "k": 8, "rotate": False, "sp": ((1, [40, 33]),)}
+        )
+
+    # The shapes DeepSeek-V4-Flash actually uses, at a context long enough that the top-k
+    # binds.  S=64 is above the operator's fused-scorer cutoff, so it is what covers the
+    # cuBLAS scoring path; the shorter steps take the fused one.
```
[the remark this answers is not in the document]
Comment (tianleiwu) [reply onnxruntime/test/python/transformers/test_sparse_row_select.py:298]: Fixed in 7257b0d210. The test sets ORT_LIGHTNING_INDEXER_MMA_SCORE=2 before session creation, forcing legal short BF16 parity cases through WMMA on SM80+; Python compile and lint pass, with CUDA runtime CI pending.
